src/services/webdav/backend.rs

```diff
@@ -274,23 +272,72 @@ impl Debug for WebdavBackend {
 impl Accessor for WebdavBackend {
     type Reader = IncomingAsyncBody;
     type BlockingReader = ();
-    type Pager = ();
+    type Pager = DirStream;
     type BlockingPager = ();
 
     fn metadata(&self) -> AccessorMetadata {
         let mut ma = AccessorMetadata::default();
         ma.set_scheme(Scheme::Webdav)
             .set_root(&self.root)
-            .set_capabilities(AccessorCapability::Read | AccessorCapability::Write)
+            .set_capabilities(
+                AccessorCapability::Read | AccessorCapability::Write | AccessorCapability::List,
+            )
             .set_hints(AccessorHint::ReadStreamable);
 
         ma
     }
 
-    async fn create(&self, path: &str, _: OpCreate) -> Result<RpCreate> {
+    async fn list(&self, path: &str, args: OpList) -> Result<(RpList, Self::Pager)> {
+        // XML body must start without a new line. Otherwise, the server will panic: `xmlParseChunk() failed`
+        let all_prop_xml_body = r#"<?xml version="1.0" encoding="utf-8" ?>
+            <D:propfind xmlns:D="DAV:">
+                <D:allprop/>
+            </D:propfind>
+        "#;
+
+        let async_body = AsyncBody::Bytes(bytes::Bytes::from(all_prop_xml_body));
         let resp = self
-            .webdav_put(path, Some(0), None, None, AsyncBody::Empty)
+            .webdav_propfind(path, None, "application/xml".into(), async_body)
             .await?;
+        let status = resp.status();
+
+        match status {
+            StatusCode::OK | StatusCode::MULTI_STATUS => {
+                let bs = resp.into_body().bytes().await?;
+                let result: Multistatus =
+                    quick_xml::de::from_reader(bs.reader()).map_err(|err| {
+                        Error::new(ErrorKind::Unexpected, &err.to_string())
+                            .with_context("service", Scheme::Webdav)
+                    })?;
+
+                Ok((
+                    RpList::default(),
+                    DirStream::new(&self.root, path, result, args.limit()),
+                ))
+            }
+            StatusCode::NOT_FOUND if path.ends_with('/') => Ok((
+                RpList::default(),
+                DirStream::new(
+                    &self.root,
+                    path,
+                    Multistatus {
+                        response: Vec::new(),
+                    },
+                    args.limit(),
+                ),
+            )),
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+
+    async fn create(&self, path: &str, _: OpCreate) -> Result<RpCreate> {
+        let resp = if path.ends_with("/") {
+            self.webdav_mkcol(path, None, None, AsyncBody::Empty)
```

Comment:
`services_webdav_list::test_list_sub_dir` won't pass unless we add MKCOL support.

There's only one test failing now.

I will continue tomorrow.

cc @Xuanwo

Reply:
Thank you so much for all you have done so far!

Reply:
All issues should be fixed.

Now directories are created recursively. Not sure whether we have better approach?

Reply:
> All issues should be fixed.

Great!

> Now directories are created recursively. Not sure whether we have better approach?

It looks good enough to get merged. We can improve this part in the future.